azure_functions_worker/dispatcher.py

```diff
@@ -334,33 +329,54 @@ async def _handle__functions_metadata_request(self, request):
             default_value=f'{PYTHON_SCRIPT_FILE_NAME_DEFAULT}')
         function_path = os.path.join(directory, script_file_name)
 
+        v2_file_exist, file_name_valid = validate_script_file(function_path,
+                                                              script_file_name)
         logger.info(
-            'Received WorkerMetadataRequest, request ID %s, function_path: %s',
-            self.request_id, function_path)
-
+            'Received WorkerMetadataRequest, request_ID %s, function_path: %s, '
+            'File_exists_and_correctly_named: %s',
+            self.request_id, function_path, v2_file_exist)
         try:
-            validate_script_file_name(script_file_name)
-
-            if not os.path.exists(function_path):
-                # Fallback to legacy model
-                logger.info("%s does not exist. "
-                            "Switching to host indexing.", script_file_name)
+            if v2_file_exist and file_name_valid:
```

Comment:
This whole if-else setup needs to be thoroughly tested.

Reply:
```
    def _dispatch_response(self, request, function_path, v2_file_exist, file_name_valid):
        conditions = [
            (v2_file_exist, file_name_valid, self._handle_valid_v2),
            (not v2_file_exist, file_name_valid, self._handle_valid_v1),
            (v2_file_exist, not file_name_valid, self._handle_invalid_v2),
            (not v2_file_exist, not file_name_valid, self._handle_invalid_app),
        ]

        for condition in conditions:
            if all(condition[:-1]):
                return condition[-1](request.request_id, function_path)

        # Default response or error handling
        return self._handle_default(request.request_id, function_path)

```
strategy design pattern may be overpowering, but a dispatch map here make code more maintainable.